documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md

```diff
@@ -0,0 +1,277 @@
+    vendor in these functions, or use a workaround that `vite` suggests (instead
+    of `const { kebab } = require "case"` use
+    `import casePkg; const { kebab } = casePkg;`)
+  - home gallery uses `require` when importing the images dynamically. There is
+    a small number of them, so we should just move them to `static` folder (or
+    `public` in Nuxt 3) and use static `import`s.
+  - [ ] #2342 Tailwind config should be converted to TypeScript.
+- Moving functionality inside the `setup` function would make it easier to
+  handle with Vue 3/Nuxt 3:
+  - Refactor `search.vue` page: move the route watcher and `fetchMedia`
+    functions to `setup`
+- Refactor image and audio single result pages to show the skeleton page when
+  the "image"/"audio" is `null`. This is required because the way data is
+  fetched changes, and _is_ the correct way it _should_ be set up (instead of
+  not handling the case of `image` being `null` and instead relying on the error
+  page redirect)
+- [x] #2234 Replace `onBeforeRouteEnter` router guards that don't have access to
+      `pinia` stores with router `middleware` that does
+- [x] #2233 Remove `@nuxtjs/style-resources` - this module was used to
+      auto-import `scss` variables into all components. We don't use `scss`
+      anymore
+
+### Changes in Nuxt 3 branch
+
+- The first PR should update the bare minimum to get the app working with Nuxt
+  3:
+  - update `nuxt` itself and `i18n`, `tailwind` and `svg-sprite` modules (and
+    the way they are set up). Remove/comment out other modules, plugins,
+    middleware.
+  - update the Typescript config and node version.
+  - Move the functions that are called at the app set up from
+    `middleware`/`layout`s to `app.vue`.
+  - Convert the layouts and pages to use `NuxtPage` instead of `NuxtChild` and
+    `slot` instead of `Nuxt`. For other necessary changes, look up the Nuxt 3
+    docs.
+  - Use the new way of declaring `async components`
+  - Remove `Vue.set`
+  - Move `key` from items inside a `v-for` to the surrounding `template`. This
+    PR would produce an app version that can be run with the `pnpm dev` or
+    `pnpm build && pnpm start` commands, but would have significantly less
+    functionality and many console errors/warnings, so it would probably be
+    pushed using `--no-verify` git flag.
+- Set up unit tests and add `.skip` to all the tests that are failing.
+  Temporarily remove the setting that disallows error logs in the test console.
+- Correctly set up the head using `useHead` and `i18n`
+- Remove the `@nuxtjs/composition-api` imports, replacing them with the
+  corresponding `useNuxtApp` or `i18n` import. If not yet possible, comment out
+  the import and the functionality that uses it.
+- Remove all `v-on="$listeners"`, making sure that the `on` handlers are still
+  passed to the correct component/element.
+- Remove the `.native` modifiers, and convert `VLink` to use `NuxtLink`. The
```

Comment:
New `NuxtLink` is basically what our `VLink` is: it handles both internal and external links: https://nuxt.com/docs/api/components/nuxt-link , so we could actually replace `VLink` with `NuxtLink`.

When trying to migrate to Nuxt 3 locally, I had some problems with the buttons that use `as="VLink"`, they wouldn't navigate correctly. I'm not sure if it's the problem with `NuxtLink` or the overall pages setup, though.

Reply:
Ahhhh, I see, I didn't realise there was a difference between Nuxt 3 `NuxtLink` and was confused because we already used `NuxtLink` :sweat_smile: The difference might be worth clarifying, especially "so we could actually replace VLink with NuxtLink", which is awesome.